pkg/services/queryhistory/database.go

```diff
@@ -324,3 +325,39 @@ func (s QueryHistoryService) migrateQueries(ctx context.Context, user *models.Si
 
 	return len(queryHistories), len(starredQueries), nil
 }
+
+func (s QueryHistoryService) deleteStaleQueries(ctx context.Context, olderThan int64) (int, error) {
+	var rowsCount int64
+
+	err := s.SQLStore.WithDbSession(ctx, func(session *sqlstore.DBSession) error {
+		sql := `DELETE 
+			FROM query_history 
+			WHERE uid IN (
+				SELECT uid FROM (
+					SELECT uid FROM query_history
+					LEFT JOIN query_history_star
+					ON query_history_star.query_uid = query_history.uid
+					WHERE query_history_star.query_uid IS NULL
+				) AS q
+			) 	
+			AND query_history.created_at <= ?`
```

Comment:
```suggestion
		sql := `DELETE 
			FROM query_history 
			WHERE uid IN (
				SELECT uid FROM (
					SELECT uid FROM query_history
					LEFT JOIN query_history_star
					ON query_history_star.query_uid = query_history.uid
					WHERE query_history_star.query_uid IS NULL
				) AS q
			) 	
			AND query_history.created_at <= ?
			ORDER BY query_history.created_at ASC
			LIMIT 10000`
```

N.B. I have not tested so that these modifications work at all.

Just to make sure that if we have _way_ to many queries to delete, we don't delete a limitless number of rows at the same time. At my previous company, I found 10 000 to typically be a reasonable value for the number of rows to delete at once in a typical SQL database (i.e. done in less than ten seconds).

I'm not suggesting any loop in case there are more, since the next cleanup will run ten minutes later and pick up another batch. This might be a problem if the rate of growth is persistently higher than 10 000/10 seconds.

Agree with @bergquist that a upper bound is probably also reasonable, and I think I might've dissuaded you from that, but what I meant wasn't to not have an upper bound but rather to make sure that that's a soft upper bound.

Reply:
Yes, makes sense. Thanks for providing all the context. Very helpful.

Reply:
Something that I've learned from operating many Grafana instances/services is that IF we somehow allow unbounded usage... some users will use it in such way. Not because they are evil. but because they dont know better and we are not stopping them. So upper limits is always a good idea. 

It's better to deal with `Please let me store more stared queries` over `Why is the database so slow?`